### PR TITLE
Add resource request and limit to init-container in pod

### PR DESCRIPTION
**Pull request by scrosby**
## Changes proposed in this PR

- When we setup the init container, attach resources to it.

## Why are we making these changes?

When K8S is run with quotas, any container without attached resources will be rejected at submission. We don't want that, so attach resources to the init container.

### CODE REVIEW

**Comment by nsinkov** [comment]
We should refactor this into a function:

**Comment by nsinkov** [comment]
let's refactor 
`    (.putRequestsItem resources "memory" (double->quantity (* memory-multiplier computed-mem)))
    (.putLimitsItem resources "memory" (double->quantity (* memory-multiplier computed-mem)))
    (.putRequestsItem resources "cpu" (double->quantity cpus))
    (when set-container-cpu-limit?
      ; Some environments may need pods to run in the "Guaranteed"
      ; QoS, which requires limits for both memory and cpu
      (.putLimitsItem resources "cpu" (double->quantity cpus)))`

into a function and use it in the 3 places: main container, sidecar, init container
so have limits in init container too

File: scheduler/src/cook/kubernetes/api.clj [hunk 1]
```diff
@@ -900,6 +920,8 @@
           (.setVolumeMounts container (filterv some? (concat [(init-container-workdir-volume-mount-fn false)
                                                               (scratch-space-volume-mount-fn false)]
                                                              init-container-checkpoint-volume-mounts)))
+          (set-mem-cpu-resources resources sidecar-mem sidecar-mem sidecar-cpu sidecar-cpu)
```
Comment (nsinkov) [inline scheduler/src/cook/kubernetes/api.clj:923]: use the total-memory-request and total-memory-limit from above

File: scheduler/src/cook/kubernetes/api.clj [hunk 2]
```diff
@@ -890,7 +899,18 @@
     ; init container
     (when use-cook-init?
       (when-let [{:keys [command image]} init-container]
-        (let [container (V1Container.)]
+        (let [container (V1Container.)
+              sidecar-mem (+ computed-mem
```
Comment (nsinkov) [inline scheduler/src/cook/kubernetes/api.clj:903]: call this total-memory-request and total-memory-limit and add those separately

Comment (scrosby) [reply scheduler/src/cook/kubernetes/api.clj]: Updated.

File: scheduler/src/cook/kubernetes/api.clj [hunk 3]
```diff
@@ -762,6 +763,19 @@
            pod-labels-from-job-labels
            pod-labels-from-job-application)))
 
+(defn set-mem-cpu-resources
+  "Given a resources object and a CPU and memory request and limit, update the resources object to reflect the
+  desired requests and limits."
+  [resources memory-request memory-limit cpu-request cpu-limit]
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/api.clj:769]: Can we add a type hint for `resources`?

File: scheduler/src/cook/kubernetes/api.clj [hunk 4]
```diff
@@ -890,7 +899,24 @@
...
-        (let [container (V1Container.)]
+        (let [container (V1Container.)
+              total-memory-request (+ computed-mem
+                                      (if use-cook-sidecar?
+                                        (get-in sidecar [:resource-requirements :memory-request])
+                                        0))
+              total-memory-limit (+ computed-mem
+                                    (if use-cook-sidecar?
+                                      (get-in sidecar [:resource-requirements :memory-limit])
+                                      0))
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/api.clj:910]: Shouldn't we use `add-as-decimals` here too?

File: scheduler/src/cook/kubernetes/api.clj [hunk 5]
```diff
@@ -890,7 +899,24 @@
...
+              total-cpu-request (add-as-decimals cpus
+                                                 (if use-cook-sidecar?
+                                                   (get-in sidecar [:resource-requirements :cpu-request])
+                                                   0))
+              total-cpu-limit (add-as-decimals cpus
+                                               (if use-cook-sidecar?
+                                                 (get-in sidecar [:resource-requirements :cpu-limit])
+                                                 0))
```
Comment (dposada) [inline scheduler/src/cook/kubernetes/api.clj:918]: Can we define a function in the `let` for this instead of repeating the code 4 times?

**Review by dposada — APPROVED** [review]

**Pull request merged**